### PR TITLE
Callbacks Refactor [3/n]: Plug in new callbacks mechanism to chat models

### PR DESCRIPTION
**Pull request by agola11**
* realized that callbacks were in the wrong place (in generate_prompt vs generate) so fixed that too.

### CODE REVIEW

File: langchain/chat_models/base.py [hunk 1]
```diff
@@ -26,99 +35,128 @@ def _get_verbosity() -> bool:
...
+
+        callback_manager = CallbackManager.configure(
+            callbacks, self.callbacks, self.verbose
+        )
+        message_strings = [get_buffer_string(m) for m in messages]
+        run_manager = callback_manager.on_llm_start(
+            {"name": self.__class__.__name__}, message_strings
+        )
+
+        new_arg_supported = inspect.signature(self._generate).parameters.get(
```
Comment (agola11) [inline langchain/chat_models/base.py:77]: This is needed for backwards compat... I'll change all of the `ChatModel` implementations in `langchain` to use the new args but others might have implemented their own `ChatModel`s .. in which case passing in an extra argument (regardless of whether or not it's optional) will break

File: langchain/callbacks/__init__.py [hunk 2]
```diff
@@ -18,51 +18,11 @@
...
-        except Exception:
-            raise ValueError(f"session {session_name} not found")
-
-
 @contextmanager
 def get_openai_callback() -> Generator[OpenAICallbackHandler, None, None]:
     """Get OpenAI callback handler in a context manager."""
     handler = OpenAICallbackHandler()
-    manager = get_callback_manager()
+    manager = CallbackManager([])
```
Comment (nfcampos) [inline langchain/callbacks/__init__.py:25]: why don't we pass the handler in the constructor?

File: langchain/callbacks/manager.py [hunk 3]
```diff
@@ -629,11 +631,10 @@ def _configure(
         )
         [callback_manager.add_handler(handler, False) for handler in local_handlers_]
 
+    if not callback_manager:
```
Comment (nfcampos) [inline langchain/callbacks/manager.py:634]: this change is a deviation from JS, where callback manager remains optional. Should I do this there too?

File: langchain/chat_models/base.py [hunk 4]
```diff
@@ -26,99 +35,128 @@ def _get_verbosity() -> bool:
...
+        self,
+        messages: List[List[BaseMessage]],
+        stop: Optional[List[str]] = None,
+        callbacks: Callbacks = None,
     ) -> LLMResult:
         """Top Level call"""
-        results = await asyncio.gather(
-            *[self._agenerate(m, stop=stop) for m in messages]
+
+        callback_manager = AsyncCallbackManager.configure(
```
Comment (nfcampos) [inline langchain/chat_models/base.py:104]: We're sure this can remain a sync method? In JS this is async

File: langchain/chat_models/base.py [hunk 5]
```diff
@@ -26,99 +35,128 @@ def _get_verbosity() -> bool:
...
-        return output
+        return await self.agenerate(prompt_messages, stop=stop)
 
     @abstractmethod
     def _generate(
-        self, messages: List[BaseMessage], stop: Optional[List[str]] = None
+        self,
+        messages: List[BaseMessage],
+        stop: Optional[List[str]] = None,
+        run_manager: Optional[CallbackManagerForLLMRun] = None,
```
Comment (nfcampos) [inline langchain/chat_models/base.py:150]: But isn't this a "breaking change" (sort of) for people who have created subclasses? After this their subclasses are invalid as they do not implement an abstract method? (They instead implement a method which doesn't exist in the super class, which by itself would be fine)

**Pull request merged**